### DISCUSSION TITLE
Docker image - ignoring configuration

### DISCUSSION THREAD

**Discussion by stephematician** [posativ/isso#1042] [Q&A] [answered]
I was experimenting with __isso__ on my local machine using __docker__ (rootless, but I believe this is irrelevant), but I was not able to change the configuration when invoking the container.

## Steps

I was able to get a comment box to appear using the following:

-   __docker__ command:
    ```
    docker run -d --rm --name isso -p 127.0.0.1:8080:8080 \
        -v /home/username/isso/config:/config \
        -v /home/username/isso/db:/db \
        ghcr.io/isso-comments/isso:release
    ```
    
-   config in `/home/username/isso/config/isso.cfg`:
    ```ini
    [general]
    dbpath = /db/comments.db
    host =
        http://localhost:1313/
    ```
    
-   and `localhost:1313` being served by __Hugo__ with `<script>` tags as per __isso__ instructions.

However, I do not believe it was using the configuration I expect. For example, if I tried to use port to 8181 via:

-   updated __docker__ command:
    ```
    docker run -d --rm --name isso -p 127.0.0.1:8181:8181 \
        -v /home/username/isso/config:/config \
        -v /home/username/isso/db:/db \
        ghcr.io/isso-comments/isso:release
    ```
    
-   updated config:
    ```ini
    [general]
    dbpath = /db/comments.db
    host =
        http://localhost:1313/
    [server]
    listen = http://0.0.0.0:8181
    ```
    
-   and I updated the `<script>` tags to refer to port 8181.

The comment box doesn't appear (and  I get __CONNECTION RESET__ errors if i try to navigate to `127.0.0.1:8181/js/embed.js`).

I can see that the __isso__ process that was spawned (inside the container) is not watching port 8181 via __netstat__:

```
$ docker exec -it isso sh
# netstat -tlunp
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name    
tcp        0      0 0.0.0.0:8080            0.0.0.0:*               LISTEN      1/python3
```

If I try running __isso__ manually within the container, then it does use the configuration I expect, e.g.:

```
$ docker exec -it isso sh
# source bin/activate
/isso # isso run &
/isso # 2024-10-16 23:34:45,199 INFO: Using configuration file '/config/isso.cfg'
2024-10-16 23:34:45,218 INFO: Using database at '/db/comments.db'
2024-10-16 23:34:45,220 WARNING: unable to connect to your website, Isso will probably not work correctly. Please make sure, Isso can reach your website via HTTP(S).
```

It is definitely listening to the port now, e.g. `netstat`.

```
/isso # netstat -tlunp
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name    
tcp        0      0 0.0.0.0:8181            0.0.0.0:*               LISTEN      30/python3
tcp        0      0 0.0.0.0:8080            0.0.0.0:*               LISTEN      1/python3
```

And the comment box is appearing.

Am I doing something wrong when I invoke docker? I tried argument `-e ISSO_SETTINGS=/config/isso.cfg` but it made no difference.

Also, if anyone has any hints about the WARNING that isso is giving me, that would be helpful <- I believe this is due to rootless docker, which might be tricky to overcome.

**Comment by stephematician** [comment]
Looks like the port is hard-coded in the Dockerfile: https://github.com/isso-comments/isso/blob/9c24f9b3d9f259cff84b67e50875e6137b459ef7/Dockerfile

Presumably the way to change the port is to build my own docker image (TBC).

**Reply by BBaoVanC** [reply, marked as answer]
You shouldn't build a custom docker image just to change the port; instead do something like this in the config:

```
[general]
dbpath = /db/comments.db
host =
    http://localhost:1313/
[server]
# this is to match what's hardcoded in the ENTRYPOINT in the Dockerfile
listen = http://0.0.0.0:8080
# this tells isso what the actual URL is
public-endpoint = http://127.0.0.1:8181
```

And in your docker command, map port 8181 on the host onto 8080 inside the container, like this: `-p 127.0.0.1:8181:8080`. This should be enough to make isso accessible at http://127.0.0.1:8181. If you want a different URL then just make sure to also change `public-endpoint` to match.

**Reply by stephematician** [reply]
Yeah, that makes sense as a solution - i.e. the ENTRYPOINT is hard coded, just leave it be, but we can map a port to that entry point with docker.

**Reply by BBaoVanC** [reply]
Yeah although technically you can change the entrypoint with the `--entrypoint` flag, but it's less neat than just updating the config

**Reply by stephematician** [reply]
Do you happen to know anything about how to get this working with rootless Docker btw?

I'm finding that from within the container, I can't contact the website that is on the host machine. This is obviously set up this way for security reasons, I've tried setting `DOCKERD_ROOTLESS_ROOTLESSKIT_DISABLE_HOST_LOOPBACK=false` as per [StackOverflow](https://stackoverflow.com/questions/72500740/how-to-access-localhost-on-rootless-docker) without success.
